candle-nn/src/encoding.rs

```diff
@@ -0,0 +1,555 @@
+                .flat_map(|i| (0..dim2).flat_map(move |j| (0..dim3).map(move |k| (i, j, k))));
+
+            match dtype {
+                DType::U8 => {
+                    let index = indices.to_vec3::<u8>()?;
+                    for (i, j, k) in iter {
+                        set_uint_value(
+                            index[i][j][k] as usize,
+                            idx(i, j, k, depth, dim2, dim3),
+                            depth,
+                            &mut v,
+                            on_value,
+                        )?;
+                    }
+                }
+                DType::U32 => {
+                    let index = indices.to_vec3::<u32>()?;
+                    for (i, j, k) in iter {
+                        set_uint_value(
+                            index[i][j][k] as usize,
+                            idx(i, j, k, depth, dim2, dim3),
+                            depth,
+                            &mut v,
+                            on_value,
+                        )?;
+                    }
+                }
+                DType::I64 => {
+                    let index = indices.to_vec3::<i64>()?;
+                    for (i, j, k) in iter {
+                        set_int_value(
+                            index[i][j][k],
+                            idx(i, j, k, depth, dim2, dim3),
+                            depth,
+                            &mut v,
+                            on_value,
+                        )?;
+                    }
+                }
+                d => unsupported_dtype(d)?,
+            };
+
+            Tensor::new(v, indices.device())?.reshape(&[dim1, dim2, dim3, depth])
+        }
+        _ => {
+            bail!("one_hot: rank {} is not supported. ", rank)
+        }
+    }
+}
+
+// Bail if the value is less than -1.
+fn check_negative(value: i64) -> Result<()> {
+    if value < -1 {
+        bail!("{}. Received {}", INVALID_ONE_HOT_INDEX_MSG, value);
+    }
+    Ok(())
+}
+
+// Bail if the value is greater than the depth.
+fn check_value_bounds(value: usize, depth: usize) -> Result<()> {
+    if value >= depth {
+        bail!(
+            "{} Index value, {}, must be less than the depth value, {}.",
+            INDEX_EXCEEDS_DEPTH_MSG,
+            value,
+            depth
+        )
+    }
+
+    Ok(())
+}
+
+// Bail if the index is out of bounds.
+fn check_idx(idx: usize, len: usize) -> Result<()> {
+    if idx >= len {
+        bail!(
+            "{} Expected index value, {}, to be less than {}.",
+            INDEX_OUT_OF_BOUNDS_MSG,
+            idx,
+            len
+        );
+    }
+
+    Ok(())
+}
+
+// Bail if the data type is not supported.
+fn unsupported_dtype(dtype: DType) -> Result<()> {
+    bail!(
+        "one_hot: Unsupported data type. Expected U8, U32, or I64. Received {:?}",
+        dtype
+    );
+}
+
+// Set unsigned integer index values to the given value.
+fn set_uint_value<D: WithDType>(
```

Comment:
This seems to be a special case of `set_int_value` below. Could we simplify things by just keeping the second?